From be16c4ed4443edcbd2e0b33602c24f4a64a6a36a Mon Sep 17 00:00:00 2001
From: Ansh Manojkumar Patel <amanojku@qti.qualcomm.com>
Date: Tue, 17 Jun 2025 15:23:46 +0530
Subject: [PATCH] PENDING: boot_temp_check feature added for u-boot

Boot Temp Check feature added for uboot.
To support the feature, new thermal driver and dtsi node is
created. The BootTempCheck functions reads the temp at bootup
and ensures if its within specified limits. If not, power off
is triggered.

Signed-off-by: Ansh Manojkumar Patel <amanojku@qti.qualcomm.com>
Upstream-Status: Pending
---
 arch/arm/dts/qcs9100-ride-r3-u-boot.dtsi |  21 +++
 arch/arm/mach-snapdragon/board.c         |  34 ++++-
 configs/qcs9100_defconfig                |   2 +
 drivers/thermal/Kconfig                  |   8 +
 drivers/thermal/Makefile                 |   1 +
 drivers/thermal/qcom_tsens.c             | 184 +++++++++++++++++++++++
 6 files changed, 249 insertions(+), 1 deletion(-)
 create mode 100644 drivers/thermal/qcom_tsens.c

diff --git a/arch/arm/dts/qcs9100-ride-r3-u-boot.dtsi b/arch/arm/dts/qcs9100-ride-r3-u-boot.dtsi
index 5c58210d96d..1cbb999d0c7 100644
--- a/arch/arm/dts/qcs9100-ride-r3-u-boot.dtsi
+++ b/arch/arm/dts/qcs9100-ride-r3-u-boot.dtsi
@@ -3,6 +3,14 @@
  * Copyright (c) 2025, Qualcomm Innovation Center, Inc. All rights reserved.
  */

+/*
+ * change the reg property of below node to change the controller base address
+ * controller_0 :  0x0C263000 (default)
+ * controller_1 :  0x0C265000
+ * controller_2 :  0x0C251000
+ * controller_3 :  0x0C252000
+ */
+
 / {
 	/* public key info */
 	signature {
@@ -109,7 +117,20 @@
 			};
 		};
 	};
+
+	btmconfig {
+		reg =<0x0C263000>;
+		#address-cells = <1>;
+		#size-cells = <0>;
+		sensor-id = <0>;
+		upper-threshold = <110>;
+		lower-threshold = <0xFFFFFFD8>;
+		maxNumIterations = <40>;
+		waitPerIterationMicroSec = <500000>;
+		compatible = "qcom,qcom_tsens";
+	};
 };
+
 / {
 	aliases {
 		/delete-property/ ethernet0;
diff --git a/arch/arm/mach-snapdragon/board.c b/arch/arm/mach-snapdragon/board.c
index 394af747ed6..585bc2b38ae 100644
--- a/arch/arm/mach-snapdragon/board.c
+++ b/arch/arm/mach-snapdragon/board.c
@@ -32,7 +32,7 @@
 #include <usb.h>
 #include <sort.h>
 #include <time.h>
-
+#include <thermal.h>
 #include "qcom-priv.h"
 
 DECLARE_GLOBAL_DATA_PTR;
@@ -427,6 +427,31 @@ void __weak qcom_late_init(void)
 
 #define addr_alloc(size) lmb_alloc(size, SZ_2M)
 
+
+// Checks the temperature at the boot up, poweroff if not within the limits.
+
+int BootTempCheck(void)
+{
+	struct udevice *dev;
+	int ret;
+	int temperature = 0;
+
+	ret = uclass_get_device_by_name(UCLASS_THERMAL, "btmconfig", &dev);
+	if (ret) {
+		debug("Failed to get thermal device with error: %d\n", ret);
+		return ret;
+	}
+
+	ret = thermal_get_temp(dev, &temperature);
+	if (ret) {
+		debug("Failed while fetching temperature with error : %d\n", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+
 /* Stolen from arch/arm/mach-apple/board.c */
 int board_late_init(void)
 {
@@ -460,6 +485,13 @@ int board_late_init(void)
 	/* Configure the dfu_string for capsule updates */
 	qcom_configure_capsule_updates();
 
+if (IS_ENABLED(CONFIG_QCOM_TSENS)) {
+	int ret_val = BootTempCheck()
+
+	if (ret_val != 0)
+		return ret_val;
+}
+
 	return 0;
 }
 
diff --git a/configs/qcs9100_defconfig b/configs/qcs9100_defconfig
index d5a22b73600..06b4d669a3e 100644
--- a/configs/qcs9100_defconfig
+++ b/configs/qcs9100_defconfig
@@ -22,6 +22,8 @@ CONFIG_SYS_LOAD_ADDR=0xD3500000
 CONFIG_CMD_BOOTQLI=y
 
 CONFIG_DM_PCA953X=y
+CONFIG_DM_THERMAL=y
+CONFIG_QCOM_TSENS=y
 
 # CONFIG_BOOTP_BOOTPATH is not set
 # CONFIG_BOOTP_DNS is not set
diff --git a/drivers/thermal/Kconfig b/drivers/thermal/Kconfig
index 440eb64a566..12e228407f0 100644
--- a/drivers/thermal/Kconfig
+++ b/drivers/thermal/Kconfig
@@ -47,4 +47,12 @@ config TI_LM74_THERMAL
 	 Enable thermal support for the Texas Instruments LM74 chip.
 	 The driver supports reading CPU temperature.
 
+config QCOM_TSENS
+        bool "Temperature sensor driver for chip"
+        help
+          Support for Temperature sensors on qcs9100.
+          It provides necessary interface to read temperature data from the on-chip thermal sensors.
+	 Enable thermal support for the QCOM chip.
+	 The driver supports reading boot time temperature.
+
 endif # if DM_THERMAL
diff --git a/drivers/thermal/Makefile b/drivers/thermal/Makefile
index b5ab0fc221f..144ea328acc 100644
--- a/drivers/thermal/Makefile
+++ b/drivers/thermal/Makefile
@@ -10,3 +10,4 @@ obj-$(CONFIG_IMX_SCU_THERMAL) += imx_scu_thermal.o
 obj-$(CONFIG_TI_DRA7_THERMAL) += ti-bandgap.o
 obj-$(CONFIG_IMX_TMU) += imx_tmu.o
 obj-$(CONFIG_TI_LM74_THERMAL) += ti-lm74.o
+obj-$(CONFIG_QCOM_TSENS) += qcom_tsens.o
diff --git a/drivers/thermal/qcom_tsens.c b/drivers/thermal/qcom_tsens.c
new file mode 100644
index 00000000000..cf70c685134
--- /dev/null
+++ b/drivers/thermal/qcom_tsens.c
@@ -0,0 +1,184 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (c) 2025, Qualcomm Innovation Center, Inc. All rights reserved.
+ */
+#include <dm.h>
+#include <thermal.h>
+#include <asm/io.h>
+#include <dm/device_compat.h>
+#include <linux/psci.h>
+#include<linux/delay.h>
+#include <log.h>
+
+#define SENSOR_OFFSET 0xA0
+#define TEMP_MASK 0xFFFF
+#define TEMP_SHIFT 0
+#define VALID_MASK 0x200000
+#define VALID_SHIFT 21
+
+
+struct tsens_priv {
+	void __iomem *base;
+	u32 offset;
+	u32 sensor_id;
+	s32 upper_thresh;
+	s32 lower_thresh;
+	u32 maxNumIterations;
+	u32 waitPerIterationMicroSec;
+};
+
+static u32 uLoopCount;
+
+static int read_prev_temp(struct udevice *dev, int *DeciDeg)
+{
+	struct tsens_priv *priv = dev_get_priv(dev);
+	u32 TsensStatus;
+	u32 validBit;
+	s32 nDeciDegC;
+
+	TsensStatus = readl(priv->base + priv->offset);
+	nDeciDegC = ((TsensStatus & TEMP_MASK) >> TEMP_SHIFT);
+	validBit = ((TsensStatus & VALID_MASK) >> VALID_SHIFT);
+
+	if (nDeciDegC & 0x800)
+		nDeciDegC |= ~0xFFF;
+
+	*DeciDeg = nDeciDegC;
+
+	if (validBit == 1)
+		return 1;
+
+	return 0;
+}
+
+static int read_temperature(struct udevice *dev, int *DeciDeg)
+{
+	s32 nDeciDegC = 0;
+	s32 nDeciDegCtry1, nDeciDegCtry2, nDeciDegCtry3;
+
+	int bValid;
+
+	bValid = read_prev_temp(dev, &nDeciDegCtry1);
+	if (bValid) {
+		nDeciDegC = nDeciDegCtry1;
+	} else {
+		bValid = read_prev_temp(dev, &nDeciDegCtry2);
+		if (bValid) {
+			nDeciDegC = nDeciDegCtry2;
+		} else {
+			bValid = read_prev_temp(dev, &nDeciDegCtry3);
+			if (bValid)
+				nDeciDegC = nDeciDegCtry3;
+			else if (nDeciDegCtry1 == nDeciDegCtry2)
+				nDeciDegC = nDeciDegCtry1;
+			else if (nDeciDegCtry2 == nDeciDegCtry3)
+				nDeciDegC = nDeciDegCtry2;
+			else
+				nDeciDegC = nDeciDegCtry1;
+		}
+	}
+
+	*DeciDeg = nDeciDegC;
+	return 0;
+}
+
+static int tsens_get_temp(struct udevice *dev, int *temp)
+{
+	struct tsens_priv *priv = dev_get_priv(dev);
+
+	if (!priv) {
+		debug("Failed to get private data\n");
+		return -EINVAL;
+	}
+
+	s32 nDeciDegC, nDegC;
+	int tsensReadResult;
+
+	for (;;) {
+		tsensReadResult = read_temperature(dev, &nDeciDegC);
+		if (tsensReadResult == 0) {
+			nDegC = nDeciDegC/10;
+
+			if (nDegC > priv->upper_thresh) {
+				if (uLoopCount == priv->maxNumIterations) {
+					debug("Temperature is %d degC(too high),
+							Going for Shutdown", nDegC);
+					psci_sys_poweroff();
+					return -EINVAL;
+				}
+			} else if (nDegC < priv->lower_thresh) {
+				if (uLoopCount == priv->maxNumIterations) {
+					debug("Temperature is %d degC(too low ),
+							Going for Shutdown", nDegC);
+					psci_sys_poweroff();
+					return -EINVAL;
+				}
+			} else {
+				if (uLoopCount >= 1) {
+					*temp = nDegC;
+					debug("Temperature is %d degC ; within the Limits,
+						continuing boot process , BTM loop count is  %d",
+						nDegC, uLoopCount);
+					return 0;
+				}
+			}
+		} else {
+			return -EINVAL;
+		}
+
+		udelay(priv->waitPerIterationMicroSec);
+		uLoopCount++;
+	}
+	*temp = 1;
+	return 0;
+}
+
+static int tsens_probe(struct udevice *dev)
+{
+	struct tsens_priv *priv = dev_get_priv(dev);
+	u32 base;
+
+	if (dev_read_u32(dev, "reg", &base))
+		debug("failed fetching base\n");
+
+	priv->base = (void __iomem *)(uintptr_t)base;
+
+	if (dev_read_u32(dev, "sensor-id", &priv->sensor_id))
+		return -EINVAL;
+
+	priv->offset = SENSOR_OFFSET + (4*priv->sensor_id);
+
+	if (dev_read_s32(dev, "upper-threshold", &priv->upper_thresh))
+		priv->upper_thresh = 110;
+	if (dev_read_s32(dev, "lower-threshold", &priv->lower_thresh))
+		priv->lower_thresh = -40;
+
+	if (dev_read_u32(dev, "maxNumIterations", &priv->maxNumIterations))
+		priv->maxNumIterations = 40;
+	if (dev_read_u32(dev, "waitPerIterationMicroSec", &priv->waitPerIterationMicroSec))
+		priv->waitPerIterationMicroSec = 500000;
+
+
+	debug("BTM sensor probed: id : %u thresholds=[%d,%d]\n",
+			priv->sensor_id, priv->upper_thresh, priv->lower_thresh);
+
+	return 0;
+}
+
+static const struct dm_thermal_ops tsens_ops = {
+	.get_temp = tsens_get_temp,
+};
+
+static const struct udevice_id tsens_ids[] = {
+	{ .compatible = "qcom,qcom_tsens" },
+	{ }
+};
+
+U_BOOT_DRIVER(qcom_tsens) = {
+	.name	= "qcom_tsens",
+	.id	= UCLASS_THERMAL,
+	.ops	= &tsens_ops,
+	.of_match = tsens_ids,
+	.probe	= tsens_probe,
+	.priv_auto = sizeof(struct tsens_priv),
+};
-- 
2.25.1

